relations/src/merkle_proof.rs

```diff
@@ -0,0 +1,61 @@
+use halo2_base::{
+    gates::{GateChip, GateInstructions},
+    poseidon::hasher::PoseidonHasher,
+    utils::BigPrimeField,
+    AssignedValue, Context,
+};
+
+use crate::poseidon_consts::{RATE, T_WIDTH};
+
+#[derive(Clone, Debug)]
+pub struct MerkleProof<F: BigPrimeField, const TREE_HEIGHT: usize> {
+    pub path_shape: [bool; TREE_HEIGHT],
+    pub path: [F; TREE_HEIGHT],
+}
+
+#[derive(Clone, Debug)]
+pub struct CircuitMerkleProof<F: BigPrimeField, const TREE_HEIGHT: usize> {
+    pub path_shape: [AssignedValue<F>; TREE_HEIGHT],
+    pub path: [AssignedValue<F>; TREE_HEIGHT],
+}
+
+impl<F: BigPrimeField, const TREE_HEIGHT: usize> MerkleProof<F, TREE_HEIGHT> {
+    pub fn new(path_shape: [bool; TREE_HEIGHT], path: [F; TREE_HEIGHT]) -> Self {
+        Self { path_shape, path }
+    }
+
+    pub fn load(&self, ctx: &mut Context<F>) -> CircuitMerkleProof<F, TREE_HEIGHT> {
+        let path_shape = self
+            .path_shape
+            .map(|x| ctx.load_witness(F::from_u128(x as u128)));
+        let path = self.path.map(|x| ctx.load_witness(x));
+
+        CircuitMerkleProof { path_shape, path }
+    }
+}
+
+impl<F: BigPrimeField, const TREE_HEIGHT: usize> CircuitMerkleProof<F, TREE_HEIGHT> {
+    pub fn verify(
+        &self,
+        ctx: &mut Context<F>,
+        gate: &GateChip<F>,
+        poseidon: &mut PoseidonHasher<F, T_WIDTH, RATE>,
+        root: AssignedValue<F>,
+        leaf: AssignedValue<F>,
+    ) {
+        let mut current_node = leaf;
+
+        for i in 0..TREE_HEIGHT {
```

Comment:
so actually we are using trees of height `TREE_HEIGHT + 1`, right? because neither the leaf nor the root is not included in the path

Reply:
Added comment about this `TREE_HEIGHT`